### PR TITLE
Add new module with http4s-grpc implementation

### PR DESCRIPTION
**Pull request by sgjbryan**
Enables use of gRPC without pulling in all the Java dependencies

- Proto definitions taken from the [OLTP spec](https://github.com/open-telemetry/opentelemetry-proto/tree/v1.0.0/opentelemetry/proto)
- Transformation from trace4cats models to OLTP protos largely taken from the existing JSON-based OLTP/HTTP implementation

### CODE REVIEW

**Comment by janstenpickle** [comment]
Thanks @sgjbryan. Rather than including the proto directly, do you know if it's maven? If so, we can use ScalaPB [to add a dependency](https://scalapb.github.io/docs/third-party-protos)

**Comment by sgjbryan** [comment]
Ah good flag. It looks like there is a [maven artifact](https://mvnrepository.com/artifact/io.opentelemetry.proto/opentelemetry-proto) but only with `alpha` versions. May still be better to use `1.0.0-alpha` rather than manually including the proto files here, what do you think?

**Comment by janstenpickle** [comment]
Hmm I think we should probably avoid using the alpha versions, it's weird that it's only those published in maven though, maybe an unpstream release issue?

Anyway, I'm wondering if we should do a submodule for these so we can pin that to a particular version, assuming we can make that with the build here of course. If that ends up being too much effort, what you've done here is fine.

File: modules/opentelemetry-otlp-http4s-grpc-exporter/src/main/scala/trace4cats/opentelemetry/otlp/proto/ExportSpansRequest.scala [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+      batch.spans
+        .foldLeft(MutMap.empty[String, ListBuffer[Span]]) { (acc, span) =>
+          acc.getOrElseUpdate(span.serviceName, ListBuffer.empty) += toSpan(span); acc
+        }
+        .view
+        .map { case (service, spans) =>
+          ResourceSpans(
+            resource = Resource(attributes =
+              List(KeyValue("service.name", AnyValue(AnyValue.Value.StringValue(service)).some))
+            ).some,
```
Comment (janstenpickle) [inline modules/opentelemetry-otlp-http4s-grpc-exporter/src/main/scala/trace4cats/opentelemetry/otlp/proto/ExportSpansRequest.scala:79]: See the recently merged #260, this should probably include all of the `TraceProcess` attributes.

File: modules/opentelemetry-otlp-http4s-grpc-exporter/src/main/scala/trace4cats/opentelemetry/otlp/OpenTelemetryOtlpHttp4sGrpcSpanExporter.scala [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import cats.effect.kernel.Temporal
+import cats.syntax.all._
+import io.opentelemetry.proto.collector.trace.v1.service.TraceService
+import org.http4s.client.Client
+import org.http4s.{Header, Headers, Uri}
+import trace4cats.kernel.SpanExporter
+import trace4cats.model.Batch
+import trace4cats.opentelemetry.otlp.proto.ExportSpansRequest
+
+object OpenTelemetryOtlpHttp4sGrpcSpanExporter {
```
Comment (janstenpickle) [inline modules/opentelemetry-otlp-http4s-grpc-exporter/src/main/scala/trace4cats/opentelemetry/otlp/OpenTelemetryOtlpHttp4sGrpcSpanExporter.scala:13]: Could you also mirror the host/port constructor like [here](https://github.com/trace4cats/trace4cats-opentelemetry/blob/master/modules/opentelemetry-otlp-grpc-exporter/src/main/scala/trace4cats/opentelemetry/otlp/OpenTelemetryOtlpGrpcSpanExporter.scala#L12-L17) please?

File: modules/opentelemetry-otlp-http4s-grpc-exporter/src/main/scala/trace4cats/opentelemetry/otlp/OpenTelemetryOtlpHttp4sGrpcSpanExporter.scala [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    * @param staticHeaders
+    * @return
+    *   OTLP/gRPC exporter
+    */
+  def fromUri[F[_]: Temporal, G[_]: Foldable](
+    client: Client[F],
+    uri: Uri,
+    staticHeaders: List[Header.ToRaw] = List.empty
+  ): SpanExporter[F, G] = new SpanExporter[F, G] {
+    val service = TraceService.fromClient[F](client, uri)
```
Comment (janstenpickle) [inline modules/opentelemetry-otlp-http4s-grpc-exporter/src/main/scala/trace4cats/opentelemetry/otlp/OpenTelemetryOtlpHttp4sGrpcSpanExporter.scala:29]: Not that it matters massively, but this should probably be `private`

**Review by janstenpickle — COMMENTED** [review]
Looks good, just a couple of nitpicks!

**Comment by sgjbryan** [comment]
> Hmm I think we should probably avoid using the alpha versions, it's weird that it's only those published in maven though, maybe an unpstream release issue?

Yeah, it seems like a release issue somewhere - a lot of the otel artifacts seem to only be published with the `-alpha` suffix.

> Anyway, I'm wondering if we should do a submodule for these so we can pin that to a particular version, assuming we can make that with the build here of course. If that ends up being too much effort, what you've done here is fine.

Not sure I quite follow what you mean here. As in just a separate SBT submodule here which contains _only_ the proto files, which is then depended on by this new exporter module?

**Comment by janstenpickle** [comment]
> Not sure I quite follow what you mean here. As in just a separate SBT submodule here with a fixed version which contains only the proto files, which is then depended on by this new exporter module?

Oh I mean a git submodule to the otel protos, that way we don't have to copy/paste them into this repo

**Comment by sgjbryan** [comment]
Ah gotcha, that makes sense. All _should_ be working now:
- Using a git submodule for the proto dependencies
- Updated the GitHub workflows to checkout submodules (think this will be necessary for the build to work)
- Addressed other PR comments

**Comment by janstenpickle** [comment]
Thanks @sgjbryan, looks like the compiler fatal warnings may need to be turned off for the generate proto code 😢

**Comment by sgjbryan** [comment]
Ah, looks like `http4s-grpc` is [only cross-published for 2.13 and 3.2](https://github.com/davenverse/http4s-grpc/blob/1b18484c52978d3703c2293b4b2eaeff26035b0c/build.sbt#L17)... Let me see what I can do

File: build.sbt [hunk 4]
```diff
@@ -86,3 +88,36 @@ lazy val `opentelemetry-otlp-http-exporter` =
...
+          case Some((2, _)) => "-Wconf:src=proto/.*:silent"
+          case _ => ""
+        }
+      },
+      scalacOptions -= {
+        CrossVersion.partialVersion(scalaVersion.value) match {
+          case Some((2, _)) => ""
+          case _ => "-Wvalue-discard"
+        }
+      }
```
[the remark this answers is not in the document]
Comment (janstenpickle) [reply build.sbt:111]: Gotta love Scala 3...

**Review by janstenpickle — APPROVED** [review]
LGTM, thanks!

**Pull request merged**